File: sdks/typescript/src/canonical.ts
import { createHash } from "node:crypto";
import * as ed from "@noble/ed25519";
import canonicalize from "canonicalize";
import type { ApsEvent } from "./types.js";

const SIG_FIELDS = ["sig_alg", "sig_value_b64"] as const;

export function stripSignatureFields(
  event: Record<string, unknown>,
): Record<string, unknown> {
  const signing: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(event)) {
    if (!SIG_FIELDS.includes(key as (typeof SIG_FIELDS)[number])) {
      signing[key] = value;
    }
  }
  return signing;
}

/** SHA-256 digest per APS-1 §3.2 (JCS of signing object + domain prefix). */
export function signingDigest(
  event: Record<string, unknown>,
  keyId: string,
): Uint8Array {
  const signing = stripSignatureFields(event);
  const jcs = canonicalize(signing);
  if (jcs === undefined) {
    throw new Error("Event is not canonicalizable");
  }
  const prefixed = `APS1\n${keyId}\n${jcs}`;
  return createHash("sha256").update(prefixed, "utf8").digest();
}

export function digestHex(event: Record<string, unknown>, keyId: string): string {
  return Buffer.from(signingDigest(event, keyId)).toString("hex");
}

export async function signEvent(
  event: ApsEvent,
  options: { privateKeyB64: string; keyId: string },
): Promise<ApsEvent> {
  const privateKey = Buffer.from(options.privateKeyB64, "base64");
  if (privateKey.length !== 32) {
    throw new Error("Ed25519 private key must be 32 bytes (base64)");
  }
  const digest = signingDigest(event as Record<string, unknown>, options.keyId);
  const signature = await ed.signAsync(digest, privateKey);
  return {
    ...event,
    sig_alg: "ed25519",
    sig_value_b64: Buffer.from(signature).toString("base64"),
  };
}

export async function verifyEventSignature(
  event: ApsEvent,
  publicKeyB64: string,
): Promise<boolean> {
  if (event.sig_alg !== "ed25519" || !event.sig_value_b64) {
    return false;
  }
  const publicKey = Buffer.from(publicKeyB64, "base64");
  if (publicKey.length !== 32) {
    return false;
  }
  const digest = signingDigest(event as Record<string, unknown>, event.key_id);
  const signature = Buffer.from(event.sig_value_b64, "base64");
  return ed.verifyAsync(signature, digest, publicKey);
}

File: sdks/python/salanor_aegis/canonical.py
"""APS-1 canonical signing (JCS + Ed25519), aligned with sdks/typescript and sdks/go."""

from __future__ import annotations

import base64
import hashlib
from typing import Any, Mapping, MutableMapping

import jcs
from nacl.exceptions import BadSignatureError
from nacl.signing import SigningKey, VerifyKey

_SIG_FIELDS = frozenset({"sig_alg", "sig_value_b64"})


def strip_signature_fields(event: Mapping[str, Any]) -> dict[str, Any]:
    return {k: v for k, v in event.items() if k not in _SIG_FIELDS}


def signing_digest(event: Mapping[str, Any], key_id: str) -> bytes:
    signing = strip_signature_fields(event)
    jcs_bytes = jcs.canonicalize(signing)
    prefixed = f"APS1\n{key_id}\n{jcs_bytes.decode('utf-8')}"
    return hashlib.sha256(prefixed.encode("utf-8")).digest()


def digest_hex(event: Mapping[str, Any], key_id: str) -> str:
    return signing_digest(event, key_id).hex()


def sign_event(
    event: Mapping[str, Any],
    *,
    private_key_b64: str,
    key_id: str,
) -> dict[str, Any]:
    priv_raw = base64.b64decode(private_key_b64)
    if len(priv_raw) != 32:
        raise ValueError("Ed25519 private key must be 32 bytes (base64 seed)")
    key = SigningKey(priv_raw)
    digest = signing_digest(event, key_id)
    sig = key.sign(digest).signature
    signed: dict[str, Any] = dict(event)
    signed["sig_alg"] = "ed25519"
    signed["sig_value_b64"] = base64.b64encode(sig).decode("ascii")
    return signed


def verify_event_signature(event: Mapping[str, Any], public_key_b64: str) -> bool:
    if event.get("sig_alg") != "ed25519" or not event.get("sig_value_b64"):
        return False
    try:
        pub_raw = base64.b64decode(public_key_b64)
        if len(pub_raw) != 32:
            return False
        verify_key = VerifyKey(pub_raw)
        digest = signing_digest(event, str(event["key_id"]))
        sig = base64.b64decode(str(event["sig_value_b64"]))
        verify_key.verify(digest, sig)
        return True
    except (BadSignatureError, ValueError, KeyError):
        return False
